src/middlewared/middlewared/plugins/smb.py

```diff
@@ -110,3 +112,255 @@ class Config:
+        path_result = await self.middleware.call(
+            'datastore.query', self._config.datastore,
+            path_filters,
+            {'prefix': self._config.datastore_prefix})
+
+        if name_result:
+            verrors.add(f'{schema_name}.name',
+                        'A share with this name already exists.')
+
+        if path_result:
+            verrors.add(f'{schema_name}.path',
+                        'A share with this path already exists.')
+
+        return name
+
+    @private
+    async def extend(self, data):
+        data['hostsallow'] = data['hostsallow'].split()
+        data['hostsdeny'] = data['hostsdeny'].split()
+
+        return data
+
+    @private
+    async def compress(self, data):
+        data['hostsallow'] = ' '.join(data['hostsallow'])
+        data['hostsdeny'] = ' '.join(data['hostsdeny'])
+
+        return data
+
+    @private
+    async def done(self, default_perms, path):
+        if default_perms:
+            try:
+                (owner, group) = await self.middleware.call(
+                    'notifier.mp_get_owner', path)
+            except Exception:
+                (owner, group) = ('root', 'wheel')
+
+            await self.middleware.call(
+                'notifier.winacl_reset', path, owner, group)
+
+    @private
+    async def get_storage_tasks(self, path=None, home=False):
+        zfs_datasets = await self.middleware.call('zfs.dataset.query')
+        task_list = []
+
+        if path:
+            for ds in zfs_datasets:
+                tasks = []
+                name = ds['name']
+                mountpoint = ds['properties']['mountpoint']['parsed']
+
+                if path == mountpoint or path.startswith(f'{mountpoint}/'):
+                    if mountpoint == path:
+                        tasks = await self.middleware.call(
+                            'datastore.query', 'storage.task',
+                            [['task_filesystem', '=', name]])
+                    else:
+                        tasks = await self.middleware.call(
+                            'datastore.query','storage.task',
+                            [['task_filesystem', '=', name],
+                             ['task_recursive', '=', 'True']])
+
+                for t in tasks:
+                    task_list.append(t)
+        elif home:
+            task_list = await self.middleware.call(
+                'datastore.query','storage.task',
+                [['task_recursive', '=', 'True']])
+
+        return task_list
+
+    @private
+    async def save(self, data):
+        task = data.get('storage_task', None)
+        home = data['home']
+        path = data['path']
+
+        if not task:
+            task_list = []
+
+            if path:
+                task_list = await self.get_storage_tasks(path=path)
+            elif home:
+                task_list = await self.get_storage_tasks(home=home)
+
+            if task_list:
+                data['storage_task'] = task_list[0]
```

Comment:
Did you test setting sotrage_task? I would think you need `task_list[0]['id']` here?